<a href="https://colab.research.google.com/github/davidmkidd/UK-Supermarket-Carbon-Emissions/blob/main/UKSmktComp_Stores_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src="https://evoviz.uk/wp-content/uploads/2026/04/Food_Divider_trans2.png">

# Clean Geolytix Retailer Points

Emissions and energy use are expected increases with business 'size'. In this second Workbook, the the Geolytix Retail Points dataset is cleaned and summarised ready for graphical explortation. Data is imported and manipulated using a combination of basic R functions and functions from external libraries, *dpylr* and *tidyr*. Code syntax is explained when first used.

# Set-up

Workbooks begin with a 'set-up' section that loads required R libraries and data.


## Load Libraries
Libraries provide additional 'specialist' functionality not supported by the 'base' program. Colab provides some commonly used libraries as 'ready-to-import' packages loaded with the *library()* function.

* [dpylr](https://dplyr.tidyverse.org/) is data manipulation library.
* [tidyr](https://tidyr.tidyverse.org/) manages ta

In [ ]:
library(dplyr)
library(tidyr)

## Download and Import Data
The next step is to import the Geolytix Retail Points dataset from the Github repository where it is stored to the /content folder.

In [ ]:
# Store the address of the csv file in a variable named 'url'
url <- "https://raw.githubusercontent.com/davidmkidd/UK-Supermarket-Carbon-Emissions/refs/heads/main/geolytix_retailpoints_v34_202412.csv"

# Store the path in Colab where the file will be saved to
download_path <- "/content/geolytix_retailpoints_v34_202412.csv"

# Download the file to the destination
download.file(url, destfile = download_path, mode = "wb")

#Import the CSV file to a variable named 'store'
store <- read.csv("/content/geolytix_retailpoints_v34_202412.csv", header=TRUE, stringsAsFactors=FALSE)

That is all the set-up required for this workbook.

# Data Frame

*read.csv()* imports data in CSV formatted text file into an R *data frame*. CSV and the R data frame are both 'table' formats in which data is stored as attribute columns and record rows. Let's explore the 'store' data frame.

In [ ]:
# ncol() returns the number of data frame columns
ncol(store)

[1] 19

In [ ]:
# names() and colnames() returns a vector of column names
names(store)

[1] "id"         "retailer"   "fascia"     "store_name" "add_one"   
 [6] "add_two"    "town"       "suburb"     "postcode"   "long_wgs"  
[11] "lat_wgs"    "bng_e"      "bng_n"      "pqi"        "open_date" 
[16] "size_band"  "county"     "year"       "size_level"

Items in a vector are accessed via their position index []

In [ ]:
# First item
names(store)[1]

# Items 4 to 8
names(store)[4:8]

# Without the third item
names(store)[-3]

[1] "id"

[1] "store_name" "add_one"    "add_two"    "town"       "suburb"

[1] "id"         "retailer"   "store_name" "add_one"    "add_two"   
 [6] "town"       "suburb"     "postcode"   "long_wgs"   "lat_wgs"   
[11] "bng_e"      "bng_n"      "pqi"        "open_date"  "size_band" 
[16] "county"     "year"       "size_level"

In [ ]:
# nrow() returns the number of data frame records
# Each row is a store
nrow(store)

[1] 18840

In [ ]:
# summary() returns a summary report of columns and their contents
summary(store)

       id             retailer            fascia           store_name       
 Min.   :1.01e+09   Length:18840       Length:18840       Length:18840      
 1st Qu.:1.01e+09   Class :character   Class :character   Class :character  
 Median :1.01e+09   Mode  :character   Mode  :character   Mode  :character  
 Mean   :1.01e+09                                                           
 3rd Qu.:1.01e+09                                                           
 Max.   :1.01e+09                                                           
                                                                            
   add_one            add_two              town              suburb         
 Length:18840       Length:18840       Length:18840       Length:18840      
 Class :character   Class :character   Class :character   Class :character  
 Mode  :character   Mode  :character   Mode  :character   Mode  :character  
                                                                            

In [ ]:
# Typing the name of a data frame returns a default number of rows.
store

id,retailer,fascia,store_name,add_one,add_two,town,suburb,postcode,long_wgs,lat_wgs,bng_e,bng_n,pqi,open_date,size_band,county,year,size_level
<int>,<chr>,<chr>,<chr>,<chr>,<chr>,<chr>,<chr>,<chr>,<dbl>,<dbl>,<dbl>,<dbl>,<chr>,<int>,<chr>,<chr>,<int>,<int>
1010008812,The Co-operative Group,The Co-operative Food,Co-op Holyhead Road - Chirk,Holyhead Road,Chirk,Chirk,,LL14 5NA,-3.058428,52.94037,328966.9,338632.9,Rooftop,NA,"< 3,013 ft2 (280m2)",Clywd,NA,1
1010021060,Spar,Spar,Spar Glyn Ceiriog,Cross Stores High Street,,Glyn Ceiriog,,LL20 7EH,-3.188270,52.93246,320226.8,337889.8,Rooftop,20230499,"< 3,013 ft2 (280m2)",Clywd,2023,1
1010007877,The Co-operative Group,The Co-operative Food,Co-op Blaenau Ffestiniog,High Street,Blaenau Ffestiniog,Blaenau Ffestiniog,Blaenau Ffestiniog,LL41 3AG,-3.934357,52.99474,270269.0,345908.0,Rooftop,NA,"3,013 < 15,069 ft2 (280 < 1,400 m2)",Gwynedd,NA,2
1010016823,Spar,Spar,Spar - Cynwyd Post Office & General Store,Cynwyd Post Office & General Store,Station Road,Cynwyd,,LL21 0LA,-3.405960,52.95915,305653.5,341122.8,Rooftop,NA,"< 3,013 ft2 (280m2)",Clywd,NA,1
1010010459,Tesco,Tesco,Tesco Wrexham Superstore,Plas Kynaston Lane,Off Well Street,Cefn-Mawr,,LL14 3AT,-3.070395,52.97315,328217.1,342292.0,Rooftop,NA,"15,069 < 30,138 ft2 (1,400 < 2,800 m2)",Clywd,NA,3
1010007420,Lidl,Lidl,Lidl Newtown,Llanidloes Road,,"Newtown, Powys",,SY16 1ET,-3.319253,52.51006,310562.0,291058.9,Rooftop,NA,"3,013 < 15,069 ft2 (280 < 1,400 m2)",Powys,NA,2
1010004399,Tesco,Tesco,Tesco Welshpool Superstore,Smithfield Road,,Welshpool,,SY21 7BL,-3.141287,52.65868,322901.3,307385.1,Rooftop,NA,"15,069 < 30,138 ft2 (1,400 < 2,800 m2)",Powys,NA,3
1010022270,Morrisons,Morrisons Daily,Morrisons Daily Welshpool,Buttington Cross Enterprise Park,,Buttington,,SY21 8SL,-3.123082,52.67150,324154.9,308791.7,Rooftop,20249999,"< 3,013 ft2 (280m2)",Powys,2024,1
1010017041,Spar,Spar,Spar - Llanrheadr,The Market Square,Oswestry,Llanrhaeadr-ym-Mochnant,,SY10 0JG,-3.301328,52.82526,312411.8,326097.0,Rooftop,NA,"< 3,013 ft2 (280m2)",Powys,NA,1


In [ ]:
# head() returns a set number of the first records.
head(store, n = 5)

,id,retailer,fascia,store_name,add_one,add_two,town,suburb,postcode,long_wgs,lat_wgs,bng_e,bng_n,pqi,open_date,size_band,county,year,size_level
,<int>,<chr>,<chr>,<chr>,<chr>,<chr>,<chr>,<chr>,<chr>,<dbl>,<dbl>,<dbl>,<dbl>,<chr>,<int>,<chr>,<chr>,<int>,<int>
1,1010008812,The Co-operative Group,The Co-operative Food,Co-op Holyhead Road - Chirk,Holyhead Road,Chirk,Chirk,,LL14 5NA,-3.058428,52.94037,328966.9,338632.9,Rooftop,NA,"< 3,013 ft2 (280m2)",Clywd,NA,1
2,1010021060,Spar,Spar,Spar Glyn Ceiriog,Cross Stores High Street,,Glyn Ceiriog,,LL20 7EH,-3.188270,52.93246,320226.8,337889.8,Rooftop,20230499,"< 3,013 ft2 (280m2)",Clywd,2023,1
3,1010007877,The Co-operative Group,The Co-operative Food,Co-op Blaenau Ffestiniog,High Street,Blaenau Ffestiniog,Blaenau Ffestiniog,Blaenau Ffestiniog,LL41 3AG,-3.934357,52.99474,270269.0,345908.0,Rooftop,NA,"3,013 < 15,069 ft2 (280 < 1,400 m2)",Gwynedd,NA,2
4,1010016823,Spar,Spar,Spar - Cynwyd Post Office & General Store,Cynwyd Post Office & General Store,Station Road,Cynwyd,,LL21 0LA,-3.405960,52.95915,305653.5,341122.8,Rooftop,NA,"< 3,013 ft2 (280m2)",Clywd,NA,1
5,1010010459,Tesco,Tesco,Tesco Wrexham Superstore,Plas Kynaston Lane,Off Well Street,Cefn-Mawr,,LL14 3AT,-3.070395,52.97315,328217.1,342292.0,Rooftop,NA,"15,069 < 30,138 ft2 (1,400 < 2,800 m2)",Clywd,NA,3


In [ ]:
# tail() returns a set number of the last records.
tail(store, n = 5)

,id,retailer,fascia,store_name,add_one,add_two,town,suburb,postcode,long_wgs,lat_wgs,bng_e,bng_n,pqi,open_date,size_band,county,year,size_level
,<int>,<chr>,<chr>,<chr>,<chr>,<chr>,<chr>,<chr>,<chr>,<dbl>,<dbl>,<dbl>,<dbl>,<chr>,<int>,<chr>,<chr>,<int>,<int>
18836,1010007793,The Co-operative Group,The Co-operative Food,Co-op Sonning Common,44 Wood Lane,Sonning Common,Sonning Common,,RG4 9SL,-0.9815925,51.51740,470761.4,180322.2,Rooftop,NA,"< 3,013 ft2 (280m2)",Oxfordshire,NA,1
18837,1010021975,Cook,Cook,Cook Henley-on-Thames,1 Oxford House,,Henley-on-Thames,,RG9 1UR,-0.9036755,51.53636,476136.2,182508.9,Rooftop,NA,"< 3,013 ft2 (280m2)",Oxfordshire,NA,1
18838,1010007999,The Co-operative Group,The Co-operative Food,Co-op Didcot,202 Broadway,,Didcot,South Didcot,OX11 8RN,-1.2440674,51.60659,452446.7,190021.2,Rooftop,NA,"3,013 < 15,069 ft2 (280 < 1,400 m2)",Oxfordshire,NA,2
18839,1010021061,Spar,Spar,Spar Wantage,18 Wallingford Street,,Wantage,Wantage,OX12 8AX,-1.4245000,51.58831,439967.5,187874.2,Rooftop,20230499,"< 3,013 ft2 (280m2)",Oxfordshire,2023,1
18840,1010014305,Iceland,Iceland,Iceland Didcot,3-5 Market Place,,Didcot,South Didcot,OX11 7LE,-1.2415885,51.60649,452618.5,190011.8,Rooftop,NA,"3,013 < 15,069 ft2 (280 < 1,400 m2)",Oxfordshire,NA,2


In [ ]:
# Values in columns are accessed with the '$' character.

store$retailer

[1] "The Co-operative Group"              
    [2] "Spar"                                
    [3] "The Co-operative Group"              
    [4] "Spar"                                
    [5] "Tesco"                               
    [6] "Lidl"                                
    [7] "Tesco"                               
    [8] "Morrisons"                           
    [9] "Spar"                                
   [10] "Spar"                                
   [11] "Spar"                                
   [12] "Spar"                                
   [13] "Spar"                                
   [14] "Aldi"                                
   [15] "Aldi"                                
   [16] "Spar"                                
   [17] "Spar"                                
   [18] "Spar"                                
   [19] "Morrisons"                           
   [20] "Aldi"                                
   [21] "Spar"                                
   [22] "Spar"                                
   [23] "Spar"                                
   [24] "The Co-operative Group"              
   [25] "Budgens"                             
   [26] "Lidl"                                
   [27] "Tesco"                               
   [28] "The Co-operative Group"              
   [29] "Sainsburys"                          
   [30] "Tesco"                               
   [31] "Marks and Spencer"                   
   [32] "Marks and Spencer"                   
   [33] "Aldi"                                
   [34] "Morrisons"                           
   [35] "Tesco"                               
   [36] "Morrisons"                           
   [37] "Farmfoods"                           
   [38] "Tesco"                               
   [39] "Morrisons"                           
   [40] "Iceland"                             
   [41] "Spar"                                
   [42] "Spar"                                
   [43] "Sainsburys"                          
   [44] "Tesco"                               
   [45] "Tesco"                               
   [46] "Tesco"                               
   [47] "Farmfoods"                           
   [48] "The Co-operative Group"              
   [49] "The Co-operative Group"              
   [50] "Tesco"                               
   [51] "The Co-operative Group"              
   [52] "Spar"                                
   [53] "The Co-operative Group"              
   [54] "Morrisons"                           
   [55] "Tesco"                               
   [56] "Iceland"                             
   [57] "The Co-operative Group"              
   [58] "Spar"                                
   [59] "Tesco"                               
   [60] "Tesco"                               
   [61] "Aldi"                                
   [62] "Tesco"                               
   [63] "Spar"                                
   [64] "Aldi"                                
   [65] "The Co-operative Group"              
   [66] "Tesco"                               
   [67] "Lidl"                                
   [68] "The Co-operative Group"              
   [69] "Morrisons"                           
   [70] "Waitrose"                            
   [71] "Asda"                                
   [72] "Asda"                                
   [73] "Tesco"                               
   [74] "Spar"                                
   [75] "Heron"                               
   [76] "Sainsburys"                          
   [77] "Aldi"                                
   [78] "Spar"                                
   [79] "Marks and Spencer"                   
   [80] "Iceland"                             
   [81] "Spar"                                
   [82] "Morrisons"                           
   [83] "Sainsburys"                          
   [84] "Sainsburys"                          
   [85] "Aldi"                                
   [86] "

These functions are all 'base' R commands. Libraries provide different functioality that build on each other to perform higher-level operations with minmial consistent syntax.

# Introducing *dpylr*

*dpylr* defines a powerful syntax that provides a simpler more consistent way of working with tabular data that always returns a table, specifically a type of table called a '[tibble](https://tibble.tidyverse.org/reference/tibble.html)', which is an 'extended' data table. The details do not matter.

For example,

In [ ]:
# dpylr (count) returns the number of rows
count(store)

n
<int>
18840


This command may, however also be written as,
```
store %>%
  count()
```
'%>%' passes the result of one statement to another allowing them to be srung together to undertake complex queries and manipulations. Here the content of 'store' is passed to count.

In [ ]:
store %>%
  count()

n
<int>
18840


Emissions data is available for eleven retailers, so one of the first questions to ask is which retailers are in the Retail Points and how many stores each has?

In [ ]:
# dpylr distinct() returns unique values.

store %>% distinct(retailer)

retailer
<chr>
The Co-operative Group
Spar
Tesco
Lidl
Morrisons
Aldi
Budgens
Sainsburys
Marks and Spencer


In [ ]:
# To count the number of unique retailers chain store, distinct(), and count()
store %>%
  distinct(retailer) %>%
  count()

n
<int>
37


Counting the number of stores each retailer has and their percentage is an example of a more complex dpylr query that was 'vibe' coded by asking Gemeni to,

> *"Count the number of store rows for each retailer and calculate the percent of records for each"*

Helpfully, Gemeni added a sort statement despite not being requested.

In [ ]:
# To count the number of stores each retailer has and calculate their percentage
store %>%
  group_by(retailer) %>%
  summarise(count = n(), .groups = "drop") %>%
  mutate(percent = count / sum(count) * 100) %>%
  arrange(desc(percent))

retailer,count,percent
<chr>,<int>,<dbl>
Tesco,2909,15.440552017
The Co-operative Group,2467,13.094479830
Spar,2118,11.242038217
Morrisons,1745,9.262208068
Sainsburys,1444,7.664543524
Asda,1106,5.870488323
Marks and Spencer,1056,5.605095541
Aldi,1043,5.536093418
Lidl,1010,5.360934183


# Clean Store Area

In the Geolytix dataset, stores are assigned to one of five size classes encoded as both *size_level* and *size_band* fields.

In [ ]:
# Unique size levels
store %>%
  distinct(size_level)

# Unique size bands
store %>%
  distinct(size_band)


size_level
<int>
1
2
3
4
5
NA


size_band
<chr>
"< 3,013 ft2 (280m2)"
"3,013 < 15,069 ft2 (280 < 1,400 m2)"
"15,069 < 30,138 ft2 (1,400 < 2,800 m2)"
"30,138 ft2 > (2,800 m2)"
"Over 30,138"
NA


The base table() fuction is a useful way of getting the number of records for each combination of class pairs.

In [ ]:
# base table() function counts records by one or more variable
table(store$size_band, store$size_level, useNA = "always")

                                        
                                             1     2     3     4     5  <NA>
  < 3,013 ft2 (280m2)                    11416     0     0     0     0     0
  15,069 < 30,138 ft2 (1,400 < 2,800 m2)     0     0  1103     0     0     0
  3,013 < 15,069 ft2 (280 < 1,400 m2)        0  4889     0     0     0     0
  30,138 ft2 > (2,800 m2)                    0     0     0  1428     0     0
  Over 30,138                                0     0     0     0     3     0
  <NA>                                       0     0     0     0     0     1

The number of  size band and level records correspond, but one record is has NA values. Let's find out what it is.

In [ ]:
# is.na() finds NAs.
# select(), which limits which columns are returned
store %>%
filter(is.na(size_level)) %>%
    select(retailer, store_name, add_one, suburb, town, postcode, year, size_level, size_band)

retailer,store_name,add_one,suburb,town,postcode,year,size_level,size_band
<chr>,<chr>,<chr>,<chr>,<chr>,<chr>,<int>,<int>,<chr>
The Co-operative Group,Co-op Welling - Wrotham Road,Wrotham Road,Welling,London,DA16 1LS,2024,NA,NA


The floor area of this store was measured from OS Mastermap data supplied through [Edina Digimap](https://digimap.edina.ac.uk/) at 820m<sup>2</sup>, within *size_level* = 2.

Use dpylr *mutate()* to update 'size_level' and 'size_band'.

In [ ]:
# WARNING
# This code updates the values of ALL records with NA values.
# This may not be approriate if there were more than one NA record.
store <- store %>%
  mutate(size_level = ifelse(is.na(size_level), 2, size_level)) %>%
  mutate(size_band = ifelse(is.na(size_band), "3,013 < 15,069 ft2 (280 < 1,400 m2)", size_band))

# Clean Year
The 'year' field stores the year the store opened.

Use *table()* to count the number of stores recorded as opening each year.

In [ ]:
table(store$year, useNA = "ifany")


    1  1855  1874  1898  1950  1967  1975  1978  1985  1989  1990  1992  1993 
    1     1     1     2     1     1     1     2     1     1     1     1     2 
 1995  1997  2000  2001  2002  2004  2007  2008  2009  2010  2011  2012  2013 
    1     1     1     3     1     1     2     2     8     4    12     8     4 
 2014  2015  2016  2017  2018  2019  2020  2021  2022  2023  2024  <NA> 
   20   362   362   736   334   502   461   428   670   934  1204 12763 

There are four things to note:

1. One store is recorded as having opened in 1AD!
2. Far fewer stores are recorded as having opened between 1855 and 2014 than subsequent years.
3. Most stores do not have year recorded.
4. Store closure is not recorded.


A store opening in 1AD is unrealistic, so may be assumed an error and year changed to NA.

In [ ]:
# Update year = 1 to NA
store <- mutate(store, year = ifelse(year == 1, NA, year))

#Check
table(store$year, useNA = "ifany")


 1855  1874  1898  1950  1967  1975  1978  1985  1989  1990  1992  1993  1995 
    1     1     2     1     1     1     2     1     1     1     1     2     1 
 1997  2000  2001  2002  2004  2007  2008  2009  2010  2011  2012  2013  2014 
    1     1     3     1     1     2     2     8     4    12     8     4    20 
 2015  2016  2017  2018  2019  2020  2021  2022  2023  2024  <NA> 
  362   362   736   334   502   461   428   670   934  1204 12764 

The Retail points data set was first released in 2014. Given the difference in  stores recorded as opening, it is reasonable to assume that the opening year of all stores existing in 2014 may not have been known. If this is the case, then 2014 is the **baseline** year to which new stores are added in subsequent years.

In [ ]:
# Copy year to year_built

store$year_built <- store$year

# Update the year values equal or less than 2014 OR (pipe character '|') NA to 2014.
store <- store %>%
  mutate(year = ifelse(year <= 2014 | is.na(store$year), 2014, year))

# Check the output
table(store$retailer, store$year, useNA = "ifany")

                                      
                                       2014 2015 2016 2017 2018 2019 2020 2021
  Aldi                                  498   70   42   94   66   60   51   57
  Allendale Co-operative Society          1    0    0    0    0    0    0    0
  Amazon                                  0    0    0    0    0    0    0   11
  Asda                                  577   31    9    8    1    0    1    0
  Booths                                 22    4    0    0    0    0    0    0
  Budgens                                56    1   10   40    4   75   24    7
  Central England Co-operative          174    0    0   31    8    3    0    3
  Channel Islands Co-operative Society   20    0    1    2    1    0    0    0
  Chelmsford Star Co-operative Society   32    0    0    4    3    0    0    0
  Clydebank Co-operative                  5    0    0    1    0    0    0    0
  Coniston Co-operative Society           1    0    0    0    0    0    0    0
  Cook       

* Records with year == 2014 are stores that existed in 2014.
* Records with year >= 2015 are new stores.

# Calculate Summary Statistics
The number and area of retailer stores in each year is calculated by cumulatively summing store number and area by year, with store area approximated as the mid-value of bound store size classes and as the lower boundary of the two upper unbounded classes.

In [ ]:
# Summarise by retailer, year and size_level
store.yr <- store %>%
  count(retailer, year, size_level, sort=TRUE)

In [ ]:
# Download geolytix_retailpoints_v34_202412.csv from GitHub
url <- "https://raw.githubusercontent.com/davidmkidd/UK-Supermarket-Carbon-Emissions/refs/heads/main/geolytix_size_class.csv"
download_path <- "/content/geolytix_size_class.csv"
download.file(url, destfile = download_path, mode = "wb")

# geolytix_size_class.csv contains the area values that will be used
# for each area class when estimating retailer store area
store.size <- read.csv("/content/geolytix_size_class.csv", header=TRUE, stringsAsFactors=FALSE)
store.size

size_level,size_area_mid,method
<int>,<int>,<chr>
1,140,mid-value
2,840,mid-value
3,2100,mid-value
4,2800,min-value
5,3000,min-value


'Joining' adds the attributes from one table onto another by matching values in fields. Here, the attriributes of store.size are added to store by matching 'size_level' fields.

The dpylr join function is called *merge()*.

In [ ]:
# Join to size class attributes to store.yr
store.yr <- merge(store.yr, store.size, by = "size_level")
head(store.yr, n=2)

,size_level,retailer,year,n,size_area_mid,method
,<dbl>,<chr>,<dbl>,<int>,<int>,<chr>
1,1,Spar,2014,1908,140,mid-value
2,1,Tesco,2014,1610,140,mid-value


Floor area of stores of each size in 2014 and new stores in subsequent years is the product of the number of stores and the size class value.

In [ ]:
# Calculate estimated store area by retailer, year, and level
store.yr$area_m2 <- store.yr$n * store.yr$size_area_mid
head(store.yr, n=2)

,size_level,retailer,year,n,size_area_mid,method,area_m2
,<dbl>,<chr>,<dbl>,<int>,<int>,<chr>,<int>
1,1,Spar,2014,1908,140,mid-value,267120
2,1,Tesco,2014,1610,140,mid-value,225400


Cumulatively sum store number and estimated area by retailer and year.

In [ ]:
# Cumulative Sum of store number by year and retailer
store.yr.n <- store.yr  %>%
  group_by(retailer, year) %>%
  arrange(retailer, year) %>%
  summarise(sum_n = sum(n), .groups = "drop_last") %>%
  mutate(cum_sum_n = cumsum(sum_n))

In [ ]:
# Cumulative Sum of store area by year and retailer
store.yr.a <- store.yr  %>%
  group_by(retailer, year) %>%
  arrange(retailer, year) %>%
  summarise(sum_area_m2 = sum(area_m2), .groups = "drop_last") %>%
  mutate(cum_sum_area_m2 = cumsum(sum_area_m2))

In [ ]:
filter(store.yr.a, retailer == "Spar")

retailer,year,sum_area_m2,cum_sum_area_m2
<chr>,<dbl>,<int>,<int>
Spar,2014,267120,267120
Spar,2022,8400,275520
Spar,2023,6720,282240
Spar,2024,14280,296520


The cumulative sums are not calculated for years in which a retailer did not open any new stores, for example Asda in 2019.

# Reformat and Export

A 'base table' of retailer and year (2014-2024) is created to which values are added.

In [ ]:
# Create 'retailer.yr' of retailer and year 2014-2024
yr <- rep(seq(2014, 2024))
retailers <- distinct(store.yr, retailer)
retailer.yr <- expand.grid(yr, retailers$retailer)
colnames(retailer.yr)[1] <- "year"
colnames(retailer.yr)[2] <- "retailer"
retailer.yr<-data.frame(retailer.yr)

Join cumulative sum tables to retailer.yr on retailer and year fields.

In [ ]:
# Join cumulative sum tables

retailer.yr <- merge(retailer.yr, store.yr.n, by.x = c("retailer","year"), by.y = c("retailer","year"), all = TRUE)
retailer.yr <- merge(retailer.yr, store.yr.a, by.x = c("retailer","year"), by.y = c("retailer","year"), all = TRUE)


head(retailer.yr, n=7)

,retailer,year,sum_n,cum_sum_n,sum_area_m2,cum_sum_area_m2
,<fct>,<int>,<int>,<int>,<int>,<int>
1,Spar,2014,1908,1908,267120,267120
2,Spar,2015,NA,NA,NA,NA
3,Spar,2016,NA,NA,NA,NA
4,Spar,2017,NA,NA,NA,NA
5,Spar,2018,NA,NA,NA,NA
6,Spar,2019,NA,NA,NA,NA
7,Spar,2020,NA,NA,NA,NA


Use tidyr *fill()* to replace NAs with the value for the previous year.

In [ ]:
# Fill missing Cum Sum values
retailer.yr <- retailer.yr%>%
  arrange(retailer, year) %>%
  fill(cum_sum_n, .direction = "down")

head(retailer.yr, n=10)


,retailer,year,sum_n,cum_sum_n,sum_area_m2,cum_sum_area_m2
,<fct>,<int>,<int>,<int>,<int>,<int>
1,Spar,2014,1908,1908,267120,267120
2,Spar,2015,NA,1908,NA,NA
3,Spar,2016,NA,1908,NA,NA
4,Spar,2017,NA,1908,NA,NA
5,Spar,2018,NA,1908,NA,NA
6,Spar,2019,NA,1908,NA,NA
7,Spar,2020,NA,1908,NA,NA
8,Spar,2021,NA,1908,NA,NA
9,Spar,2022,60,1968,8400,275520


In [ ]:
retailer.yr<- retailer.yr %>%
  arrange(retailer, year) %>%
  fill(cum_sum_area_m2, .direction = "down")

In [ ]:
head(retailer.yr, n=2)

,retailer,year,sum_n,cum_sum_n,sum_area_m2,cum_sum_area_m2
,<fct>,<int>,<int>,<int>,<int>,<int>
1,Spar,2014,1908,1908,267120,267120
2,Spar,2015,NA,1908,NA,267120


In [ ]:
# Rename columns
retailer.yr <- retailer.yr %>% rename("new_store" = "sum_n",
                    "total_store" = "cum_sum_n",
                    "new_area" = "sum_area_m2",
                    "total_area" = "cum_sum_area_m2")

In [ ]:
names(retailer.yr)

[1] "retailer"    "year"        "new_store"   "total_store" "new_area"   
[6] "total_area"

The final steps deal with ...

In [ ]:
# Download retailer data
url <- "https://raw.githubusercontent.com/davidmkidd/UK-Supermarket-Carbon-Emissions/refs/heads/main/retailer_data.csv"
download_path <- "/content/retailer_data.csv"
download.file(url, destfile = download_path, mode = "wb")
retailer.data <- read.csv("/content/retailer_data.csv", header=TRUE, stringsAsFactors=FALSE)

In [ ]:
# Add Retailer Code
retailer.yr <- merge(retailer.yr, retailer.data, by = "retailer", all.x = TRUE)

In [ ]:
# Set New Store and New Area which are NA to zero except for Ocardo

retailer.yr <- retailer.yr %>%
  mutate(new_store = ifelse(retailer != "Ocado" & is.na(new_store), 0, new_store)) %>%
  mutate(new_area = ifelse(retailer != "Ocado" & is.na(new_area), 0, new_area))

In [ ]:
# Set Ocado Total Store and Area to NA

retailer.yr <- retailer.yr %>%
  mutate(total_store = ifelse(retailer == "Ocado", NA, total_store)) %>%
  mutate(total_area = ifelse(retailer == "Ocado", NA, total_area))

In [ ]:
# Check
head(retailer.yr, n = 8)

,retailer,year,new_store,total_store,new_area,total_area,retailer_code,hex,revenue_2022,revenue_2023,mean_revenue,cost_index
,<fct>,<int>,<dbl>,<int>,<dbl>,<int>,<chr>,<chr>,<chr>,<chr>,<int>,<dbl>
1,Spar,2014,1908,1908,267120,267120,NA,NA,NA,NA,NA,NA
2,Spar,2015,0,1908,0,267120,NA,NA,NA,NA,NA,NA
3,Spar,2016,0,1908,0,267120,NA,NA,NA,NA,NA,NA
4,Spar,2017,0,1908,0,267120,NA,NA,NA,NA,NA,NA
5,Spar,2018,0,1908,0,267120,NA,NA,NA,NA,NA,NA
6,Spar,2019,0,1908,0,267120,NA,NA,NA,NA,NA,NA
7,Spar,2020,0,1908,0,267120,NA,NA,NA,NA,NA,NA
8,Spar,2021,0,1908,0,267120,NA,NA,NA,NA,NA,NA


In [ ]:
retailer.yr %>%
  summarise(total_store = sum(new_store, na.rm = TRUE))

total_store
<dbl>
18840


Export cleaned store and store.yr to csv file for next tutorial.

In [ ]:
 write.csv(store,file='/content/geolytix_retailpoints_v34_202412_clean.csv', row.names=FALSE)
 write.csv(retailer.yr,file='/content/geolytix_retailpoints_v34_202412_summary.csv', row.names=FALSE)

---

Back: [UKSmktComp Main Page](https://colab.research.google.com/drive/1f8a0pXfF9PqCujiwjf4TO4-k7ezt-6b3?usp=sharing)

Next: [Visualising Retailer Stores](https://colab.research.google.com/drive/170bNbYyXBmAjgpsRyzqzruE32Jw_5NHR#scrollTo=oUcubW-WW7ok)

---